statistics/feedback.go

```diff
@@ -493,3 +493,127 @@ func buildNewHistogram(h *Histogram, buckets []bucket) *Histogram {
 	}
 	return hist
 }
+
+// queryFeedback is used to serialize the QueryFeedback.
+type queryFeedback struct {
+	IntRanges   []int64
+	HashValues  []uint64 // HashValues is the murmur hash values for each index point.
+	IndexRanges [][]byte
+	Counts      []int64 // Counts is the number of scan keys in each range.
+}
+
+func encodePKFeedback(q *QueryFeedback) (*queryFeedback, error) {
+	pb := &queryFeedback{}
+	for _, fb := range q.feedback {
+		err := q.DecodeInt()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		// There is no need to update the point queries.
+		if fb.upper.GetInt64()-fb.lower.GetInt64() <= 1 {
+			continue
+		}
+		pb.IntRanges = append(pb.IntRanges, fb.lower.GetInt64(), fb.upper.GetInt64())
+		pb.Counts = append(pb.Counts, fb.count)
+	}
+	return pb, nil
+}
+
+func encodeIndexFeedback(q *QueryFeedback) *queryFeedback {
+	pb := &queryFeedback{}
+	var pointCounts []int64
+	for _, fb := range q.feedback {
+		if bytes.Equal(kv.Key(fb.lower.GetBytes()).PrefixNext(), fb.upper.GetBytes()) {
+			h1, h2 := murmur3.Sum128(fb.lower.GetBytes())
+			pb.HashValues = append(pb.HashValues, h1, h2)
+			pointCounts = append(pointCounts, fb.count)
+		} else {
+			pb.IndexRanges = append(pb.IndexRanges, fb.lower.GetBytes(), fb.upper.GetBytes())
+			pb.Counts = append(pb.Counts, fb.count)
+		}
+	}
+	pb.Counts = append(pb.Counts, pointCounts...)
+	return pb
+}
+
+func encodeFeedback(q *QueryFeedback) ([]byte, error) {
+	var pb *queryFeedback
+	var err error
+	if q.hist.tp.Tp == mysql.TypeLong {
```

Comment:
Can we make sure that the Tp is only `TypeLong`? Does it possible to be `TypeLongLong`?